server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -190,35 +191,90 @@ protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<
+                if (executionMode == null) {
+                    boolean collapseWithInnerHits = source != null && source.collapse() != null && source.collapse().getInnerHits() != null
+                        && source.collapse().getInnerHits().isEmpty() == false;
+                    boolean scroll = searchRequest.scroll() != null;
+                    executionMode = collapseWithInnerHits || scroll ? CCSReduceMode.LOCAL
+                        : CCSReduceMode.REMOTE;
+                }
+                if (executionMode == CCSReduceMode.LOCAL) {
+                    AtomicInteger skippedClusters = new AtomicInteger(0);
+                    collectSearchShards(searchRequest.indicesOptions(), searchRequest.preference(), searchRequest.routing(),
+                        skippedClusters, remoteClusterIndices, remoteClusterService, threadPool,
+                        ActionListener.wrap(
+                            searchShardsResponses -> {
+                                List<SearchShardIterator> remoteShardIterators = new ArrayList<>();
+                                Map<String, AliasFilter> remoteAliasFilters = new HashMap<>();
+                                BiFunction<String, String, DiscoveryNode> clusterNodeLookup = processRemoteShards(
+                                    searchShardsResponses, remoteClusterIndices, remoteShardIterators, remoteAliasFilters);
+                                int localClusters = localIndices == null ? 0 : 1;
+                                int totalClusters = remoteClusterIndices.size() + localClusters;
+                                int successfulClusters = searchShardsResponses.size() + localClusters;
+                                executeSearch((SearchTask) task, timeProvider, searchRequest, localIndices,
+                                    remoteShardIterators, clusterNodeLookup, clusterState, remoteAliasFilters, listener,
+                                    new SearchResponse.Clusters(totalClusters, successfulClusters, skippedClusters.get(),
+                                        CCSReduceMode.LOCAL));
+                            },
+                            listener::onFailure));
+                } else {
+                    //TODO make this a method and test it?
+                    final int from;
+                    final int size;
+                    final int trackTotalHitsUpTo;
+                    if (source == null) {
+                        from = SearchService.DEFAULT_FROM;
+                        size = SearchService.DEFAULT_SIZE;
+                        trackTotalHitsUpTo = SearchContext.DEFAULT_TRACK_TOTAL_HITS_UP_TO;
+                    } else {
+                        from = source.from() == -1 ? SearchService.DEFAULT_FROM : source.from();
+                        size = source.size() == -1 ? SearchService.DEFAULT_SIZE : source.size();
+                        trackTotalHitsUpTo = source.trackTotalHitsUpTo();
```

Comment:
`source.trackTotalHitsUpTo()` is nullable in https://github.com/elastic/elasticsearch/pull/37466 so you'll need to check here too and replace with `SearchContext.DEFAULT_TRACK_TOTAL_HITS_UP_TO` if it's the case.

Reply:
good one, I will update once the linked PR is in.

Reply:
I have updated the PR accordingly.